Cannot pull port (requestParsingIn) twice

Environment: Java 8, scala 2.11.7, akka 2.4.8

Testing with:
curl -v -F"a=@test1.dat" -F"b=@test2.dat" http://localhost:7777/test

The following code results in a Timeout Error:

```
object AkkaTest extends App {
  implicit val system = ActorSystem()
  implicit val materializer = ActorMaterializer()
  implicit val executor = system.dispatcher

  val route = encodeResponse {
    (post & path("test") & entity(as[Multipart.FormData])) { entity =>
      complete {
        val sources = entity.parts.map(e => e.entity.dataBytes)
        val f = sources.runForeach { source =>
          val inputStream = source.runWith(StreamConverters.asInputStream())
          try {
            val tmp = Files.createTempFile("a", "b")
            Files.copy(inputStream, tmp, StandardCopyOption.REPLACE_EXISTING)
            Files.delete(tmp)
          } finally {
            inputStream.close()
          }
        }
        f.map(_ => "ok")
      }
    }
  }
  import akka.http.scaladsl.Http
  Http().bindAndHandle(route, "0.0.0.0", 7777)
```

Whereas, this works as expected (using runFold and ByteArrrayInputStream):

```
object AkkaTest extends App {
  implicit val system = ActorSystem()
  implicit val materializer = ActorMaterializer()
  implicit val executor = system.dispatcher

  val route = encodeResponse {
    (post & path("test") & entity(as[Multipart.FormData])) { entity =>
      complete {
        val sources = entity.parts.map(e => e.entity.dataBytes)
        val f = sources.runForeach { source =>
          val tmp = Files.createTempFile("a", "b")
          source.runFold(ByteString.empty)((bs, c) => bs ++ c)
            .foreach { bs =>
              val inputStream = new ByteArrayInputStream(bs.toArray)
              try {
                Files.copy(inputStream, tmp, StandardCopyOption.REPLACE_EXISTING)
                Files.delete(tmp)
              } finally {
                inputStream.close()
              }
            }
        }
        f.map(_ => "ok")
      }
    }
  }
  import akka.http.scaladsl.Http
  Http().bindAndHandle(route, "0.0.0.0", 7777)
}
```

Exception:
```
java.io.IOException: Timeout on waiting for new data
	at akka.stream.impl.io.InputStreamAdapter$$anonfun$read$1.apply$mcI$sp(InputStreamSinkStage.scala:144) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter$$anonfun$read$1.apply(InputStreamSinkStage.scala:129) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter$$anonfun$read$1.apply(InputStreamSinkStage.scala:129) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter.executeIfNotClosed(InputStreamSinkStage.scala:109) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter.read(InputStreamSinkStage.scala:129) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.io.InputStreamAdapter.read(InputStreamSinkStage.scala:120) ~[akka-stream_2.11-2.4.8.jar:na]
	at java.nio.file.Files.copy(Files.java:2908) ~[na:1.8.0_45]
	at java.nio.file.Files.copy(Files.java:3027) ~[na:1.8.0_45]
	at com.scb.sabre.store.http.AkkaTest$$anonfun$15$$anonfun$apply$69$$anonfun$apply$70$$anonfun$17.apply(
....
2016-07-20 10:02:13,202 - [ERROR] - [akka.actor.RepointableActorRef] [undefined] [default-akka.actor.default-dispatcher-2] - Error in stage [akka.http.impl.engine.server.HttpServerBluePrint$ControllerStage@5455fb41]: requirement failed: Cannot pull port (requestParsingIn) twice
java.lang.IllegalArgumentException: requirement failed: Cannot pull port (requestParsingIn) twice
	at scala.Predef$.require(Predef.scala:219) ~[scala-library-2.11.6.jar:na]
	at akka.stream.stage.GraphStageLogic.pull(GraphStage.scala:348) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.http.impl.engine.server.HttpServerBluePrint$ControllerStage$$anon$12$$anon$15.onPush(HttpServerBluePrint.scala:428) ~[akka-http-core_2.11-2.4.8.jar:na]
	at akka.stream.impl.fusing.GraphInterpreter.processElement$1(GraphInterpreter.scala:590) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:601) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:542) ~[akka-stream_2.11-2.4.8.jar:na]
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:471) ~[akka-stream_2.11-2.4.8.jar:na]

```

Possibly related to:
https://github.com/akka/akka/issues/19392
https://github.com/akka/akka/issues/20642

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala
@@ -383,7 +383,7 @@ private[http] object HttpServerBluePrint {
               push(requestPrepOut, rs)
             case MessageEnd ⇒
               messageEndPending = false
-              if(isAvailable(requestPrepOut)) push(requestPrepOut, MessageEnd)
+              push(requestPrepOut, MessageEnd)
             case MessageStartError(status, info) ⇒ finishWithIllegalRequestError(status, info)
             case x: EntityStreamError if messageEndPending && openRequests.isEmpty ⇒
               // client terminated the connection after receiving an early response to 100-continue

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala
@@ -383,7 +383,7 @@ private[http] object HttpServerBluePrint {
               push(requestPrepOut, rs)
             case MessageEnd ⇒
               messageEndPending = false
-              push(requestPrepOut, MessageEnd)
+              if(isAvailable(requestPrepOut)) push(requestPrepOut, MessageEnd)
             case MessageStartError(status, info) ⇒ finishWithIllegalRequestError(status, info)
             case x: EntityStreamError if messageEndPending && openRequests.isEmpty ⇒
               // client terminated the connection after receiving an early response to 100-continue
@@ -421,7 +421,7 @@ private[http] object HttpServerBluePrint {
 
           emit(responseCtxOut, ResponseRenderingContext(response, requestStart.method, requestStart.protocol, close),
             pullHttpResponseIn)
-          if (close && requestStart.expect100Continue) pull(requestParsingIn)
+          if (!isClosed(requestParsingIn) && close && requestStart.expect100Continue) pull(requestParsingIn)
         }
         override def onUpstreamFinish() =
           if (openRequests.isEmpty && isClosed(requestParsingIn)) completeStage()

--- a/akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerBug2108Spec.scala
+++ b/akka-http-core/src/test/scala/akka/http/impl/engine/server/HttpServerBug2108Spec.scala
@@ -1,0 +1,91 @@
+/*
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.http.impl.engine.server
+
+import akka.http.scaladsl.model.HttpEntity.Chunked
+import akka.http.scaladsl.model.HttpMethods._
+import akka.http.scaladsl.model.{ ContentType, HttpRequest, HttpResponse }
+import akka.http.scaladsl.model.MediaTypes._
+import akka.stream.{ ActorMaterializer, Materializer }
+import akka.stream.testkit.Utils.{ TE, _ }
+import akka.testkit.{ AkkaSpec, EventFilter }
+import org.scalatest.Inside
+
+import scala.concurrent.Await
+import scala.concurrent.duration._
+
+class HttpServerBug2108Spec extends AkkaSpec(
+  """
+   akka.loglevel = WARNING
+   akka.loggers = ["akka.testkit.TestEventListener", "akka.event.Logging$DefaultLogger"]
+   akka.http.server.request-timeout = infinite
+   akka.test.filter-leeway=300ms""") with Inside { spec ⇒
+  implicit val materializer = ActorMaterializer()
+
+  "The HttpServer" should {
+
+    "not cause internal graph failures when consuming a `100 Continue` entity triggers a failure" in assertAllStagesStopped(new HttpServerTestSetupBase {
+      override implicit def system = HttpServerBug2108Spec.this.system
+      override implicit def materializer: Materializer = HttpServerBug2108Spec.this.materializer
+
+      send("""POST / HTTP/1.1
+             |Host: example.com
+             |Expect: 100-continue
+             |Transfer-Encoding: chunked
+             |
+             |""")
+      inside(expectRequest()) {
+        case HttpRequest(POST, _, _, Chunked(ContentType(`application/octet-stream`, None), data), _) ⇒
+          val done = data.runForeach(_ ⇒ throw TE("failed on first chunk"))
+
+          expectResponseWithWipedDate(
+            """HTTP/1.1 100 Continue
+              |Server: akka-http/test
+              |Date: XXXX
+              |
+              |""")
+          send("""10
+                 |0123456789ABCDEF
+                 |0
+                 |
+                 |""")
+
+          // Bug #21008 does not actually ever make the request fail instead it logs the exception and behaves
+          // nicely so we need to check that the exception didn't get logged
+          var sawException = false
+          try {
+            EventFilter[IllegalArgumentException](occurrences = 1) intercept {
+              // make sure the failure has happened
+              Await.ready(done, 10.seconds)
+              // and then when the failure has happened/future completes, we push a reply
+              responses.sendNext(HttpResponse(entity = "Yeah"))
+
+            }
+            // got such an error, that is bad,
+            sawException = true
+          } catch {
+            case _: AssertionError ⇒ sawException = false
+          }
+          if (sawException) fail("HttpServerBluePrint.ControllerStage: requirement failed: Cannot pull closed port (requestParsingIn)")
+
+          // and the client should still get that ok
+          expectResponseWithWipedDate(
+            """HTTP/1.1 200 OK
+              |Server: akka-http/test
+              |Date: XXXX
+              |Connection: close
+              |Content-Type: text/plain; charset=UTF-8
+              |Content-Length: 4
+              |
+              |Yeah""")
+
+      }
+
+      netIn.sendComplete()
+      netOut.expectComplete()
+
+    })
+
+  }
+}

